tests/unit/test_basics.py

```diff
@@ -110,33 +137,127 @@ def assert_user_state_empty():
+
+    def test_studio_submit_empty_max_items(self):
+        def update_submission(submission):
+            submission['max_items_per_zone'] = ''
+
+        body = self._make_submission(update_submission)
+        res = self.call_handler('studio_submit', body)
+        self.assertEqual(res, {'result': 'success'})
+
+        self.assertIsNone(self.block.max_items_per_zone)
+
+    @ddt.data(-1, -5, -100, -1e15)
+    def test_studio_submit_max_items_negative(self, max_items_per_zone):
+        def modify_submission(submission):
+            submission['max_items_per_zone'] = max_items_per_zone
+
+        res = self.call_handler('studio_submit', self._make_submission(modify_submission))
+        self.assertEqual(res['result'], 'failure')
+        self.assertEqual(len(res['messages']), 1)
+        self.assertIn('should be positive', res['messages'][0])
+
+    def test_studio_submit_max_items_zero_fails_validation(self):
+        def modify_submission(submission):
+            submission['max_items_per_zone'] = 0
+
+        res = self.call_handler('studio_submit', self._make_submission(modify_submission))
+        self.assertEqual(res['result'], 'failure')
+        self.assertEqual(len(res['messages']), 1)
+        self.assertIn("should be positive integer", res['messages'][0])
+
+    def test_studio_submit_max_items_validation_simple(self):
+        def modify_submission(submission):
+            submission['max_items_per_zone'] = 1
+            submission['data']['items'] = [
+                {'zones': ['Zone 1'], 'title': 'item 1'}, {'zones': ['Zone 2'], 'title': 'item 2'}
+            ]
+
+        res = self.call_handler('studio_submit', self._make_submission(modify_submission))
+        self.assertEqual(res, {'result': 'success'})
+
+    def test_studio_submit_max_items_validation_multiple_items(self):
+        def modify_submission(submission):
+            submission['max_items_per_zone'] = 2
+            submission['data']['items'] = [
+                {'zones': ['Zone 1'], 'title': 'item 1'},
+                {'zones': ['Zone 1'], 'title': 'item 2'}
+            ]
+
+        res = self.call_handler('studio_submit', self._make_submission(modify_submission))
+        self.assertEqual(res, {'result': 'success'})
+
+    def test_studio_submit_max_items_validation_failure_simple(self):
+        def modify_submission(submission):
+            submission['max_items_per_zone'] = 1
+            submission['data']['items'] = [
+                {'zones': ['Zone 1'], 'title': 'item 1'}, {'zones': ['Zone 1'], 'title': 'item 2'}
+            ]
+
+        res = self.call_handler('studio_submit', self._make_submission(modify_submission))
+        self.assertEqual(res['result'], 'failure')
+        self.assertEqual(len(res['messages']), 1)
+        self.assertIn("Zone 1", res['messages'][0])
+
+    def test_studio_submit_max_items_validation_failure_multiple_items(self):
+        def submission_failure2(submission):
```

Comment:
@e-kolpakov This should be called `modify_submission` to be consistent with the other functions that change the default submission. There's also a couple of functions that serve the same purpose and are called `update_submission` (in L157 and L185 above); it would make sense to rename those as well.